runtime/binding-http/src/main/java/io/aklivity/zilla/runtime/binding/http/internal/stream/HttpClientFactory.java

```diff
@@ -856,103 +859,148 @@ int decode(
     {
         private final long resolvedId;
         private final List<HttpClient> clients;
-        private final Queue<HttpRequest> requests;
+
+        private final Long2ObjectHashMap<HttpExchange> exchanges;
+        private int headerSlot = NO_SLOT;
+        private int headerSlotOffset;
+        private int headerSlotLimit;
+        private SortedSet<HttpVersion> versions;
 
         private HttpClientPool(
             long resolvedId)
         {
             this.resolvedId = resolvedId;
             this.clients = new LinkedList<>();
-            this.requests = new LinkedList<>();
+            this.exchanges = new Long2ObjectHashMap();
         }
 
         public MessageConsumer newStream(
+            DirectBuffer buffer,
+            int index,
+            int length,
             MessageConsumer sender,
-            BeginFW begin,
-            Map<String8FW, String16FW> overrides)
+            Map<String8FW, String16FW> overrides,
+            SortedSet<HttpVersion> versions)
         {
             // count all requests
             countRequests.getAsLong();
+            this.versions = versions;
 
+            final BeginFW begin = beginRO.wrap(buffer, index, index + length);
             HttpClient client = supplyClient();
 
             MessageConsumer newStream = null;
 
-            if (client != null)
+            final int extensionLength = begin.extension().sizeof();
+            if (extensionLength > maximumHeadersSize)
             {
-                final HttpExchange exchange = client.newExchange(sender, begin, overrides);
-                newStream = exchange::onApplication;
+                HttpBeginExFW beginEx = beginExRW.wrap(codecBuffer, 0, codecBuffer.capacity())
+                        .typeId(httpTypeId)
+                        .headersItem(h -> h.name(HEADER_STATUS).value(STATUS_431))
+                        .build();
+                newStream = rejectWithStatusCode(sender, begin, beginEx);
```

Comment:
Consider moving the construction of `beginEx` inside `rejectWithStatusCode` and pass only `STATUS_431` here instead of `beginEx`.

Reply:
I did that but then in the case of 503 there is another header `retry-after` hence I moved the headers out of the method.